examples/custom_rule.jl

```diff
@@ -0,0 +1,191 @@
+# # Enzyme custom rules tutorial
+
+# The goal of this tutorial is to give a simple example of defining a custom rule with Enzyme.
+# Specifically, our goal will be to write custom rules for the following function `f`:
+
+function f(y, x)
+    y .= x.^2
+    return sum(y) 
+end
+
+# Our function `f` populates its first input `y` with the element-wise square of `x`.
+# In addition to doing this mutation, it returns `sum(y)` as output. What a sneaky function!
+
+# In this case, Enzyme can differentiate through `f` automatically. For example, using forward mode:
+
+using Enzyme
+x  = [3.0, 1.0]
+dx = [1.0, 0.0]
+y  = [0.0, 0.0]
+dy = [0.0, 0.0]
+
+g(y, x) = f(y, x)^2 # function to differentiate 
+
+@show autodiff(Forward, g, Duplicated(y, dy), Duplicated(x, dx)) # derivative of g w.r.t. x[1]
+@show dy; # derivative of y w.r.t. x[1] when g is run
+
+# (See the [AutoDiff API tutorial](autodiff.md) for more information on using `autodiff`.)
+
+# But there may be special cases where we need to write a custom rule to help Enzyme out.
+# Let's see how to write a custom rule for `f`!
+
+# !!! warning "Don't use custom rules unnecessarily!"
+#     Enzyme can efficiently handle a wide range of constructs, and so a custom rule should only be required
+#     in certain special cases. For example, a function may make a foreign call that Enzyme cannot differentiate,
+#     or we may have higher-level mathematical knowledge that enables us to write a more efficient rule. 
+#     Even in these cases, try to make your custom rule encapsulate the minimum possible construct that Enzyme
+#     cannot differentiate, rather than expanding the scope of the rule unnecessarily.
+#     For pedagogical purposes, we will disregard this principle here and go ahead and write a custom rule for `f` :)
+
+# ## Defining our first (forward-mode) rule 
+
+# First, we import the functions [`EnzymeRules.forward`](@ref), [`EnzymeRules.augmented_primal`](@ref),
+# and [`EnzymeRules.reverse`](@ref).
+# We need to overload `forward` in order to define a custom forward rule, and we need to overload
+# `augmented_primal` and `reverse` in order to define a custom reverse rule.
+
+import .EnzymeRules: forward, reverse, augmented_primal
+using .EnzymeRules
+
+# In this section, we write a simple forward rule to start out:
+
+function forward(func::Const{typeof(f)}, ::Type{<:Duplicated}, y::Duplicated, x::Duplicated)
+    println("Using custom rule!")
+    out = func.val(y.val, x.val)
+    y.dval .= 2 .* x.val .* x.dval
```

Comment:
is using `x.val` here incorrect, because it could have been mutated since? Do I need to use the tape if I want the rule to handle possible overwrites?